sql/src/main/java/io/crate/planner/consumer/QueryAndFetchConsumer.java

```diff
@@ -99,25 +102,113 @@ public Plan visitQueriedTable(QueriedTable table, ConsumerContext context) {
-            if (qs.hasAggregates() || qs.groupBy().isPresent()) {
+            if (!isSimpleSelect(qs, context)) {
                 return null;
             }
             Planner.Context plannerContext = context.plannerContext();
-            Plan plan = Merge.ensureOnHandler(
-                plannerContext.planSubRelation(relation.subRelation(), context), plannerContext);
-
-            Limits limits = plannerContext.getLimits(qs);
-            maybeAddFilterProjection(relation, plan);
-            Projection topN = ProjectionBuilder.topNOrEval(
-                relation.subRelation().fields(),
-                qs.orderBy().orElse(null),
+            QueriedRelation subRelation = relation.subRelation();
+            FetchDecider parentFetchDecider = context.fetchDecider();
+            context.setFetchDecider(FetchDecider.NO_FINALIZE);
+            Plan subPlan = plannerContext.planSubRelation(subRelation, context);
+            context.setFetchDecider(parentFetchDecider);
+
+            FetchRewriter.FetchDescription fetchDescription = subPlan.resultDescription().fetchDescription();
+            if (fetchDescription == null) {
+                return finalizeQueriedSelectPlanWithoutFetch(qs, plannerContext, subRelation, subPlan);
+            }
+            return tryToPropagateFetch(qs, subRelation, subPlan, context);
+        }
+    }
+
+    private static Plan tryToPropagateFetch(QuerySpec querySpec,
+                                            QueriedRelation subRelation,
+                                            Plan plan,
+                                            ConsumerContext context) {
+        FetchRewriter.FetchDescription fetchDescription = plan.resultDescription().fetchDescription();
+        Planner.Context plannerContext = context.plannerContext();
+        plan = Merge.ensureOnHandler(plan, plannerContext);
+
+        WhereClause where = querySpec.where();
+        boolean appliedWhere = tryApplyWherePreFetch(plan, fetchDescription, where);
+
+        Limits limits = plannerContext.getLimits(querySpec);
+        boolean appliedOrderByAndLimit = tryApplyOrderAndLimitPreFetch(querySpec, plan, fetchDescription, limits);
+
+        if (appliedWhere && appliedOrderByAndLimit && !context.fetchDecider().finalizeFetch()) {
+            fetchDescription.updatePostFetchOutputs(querySpec.outputs());
+            return new PlanWithFetchDescription(plan, fetchDescription);
+        }
+
+        plan = planFetch(fetchDescription, plannerContext, plan);
+        if (!appliedWhere) {
+            applyFilter(plan, subRelation.fields(), where);
+        }
+        if (appliedOrderByAndLimit) {
+            // TODO: should we try to skip evalProjection by checking if there are any functions/column-re-orderings?
+            EvalProjection eval = new EvalProjection(
+                InputColumns.create(querySpec.outputs(), new InputColumns.Context(subRelation.fields())));
+            plan.addProjection(eval, null, null, null);
```

Comment:
Yep I agree, thought about that too but I'd like to look into this in a separate PR